supply/lib/supply/client.rb

```diff
@@ -5,7 +5,67 @@
 require 'net/http'
 
 module Supply
-  class Client
+  UI = FastlaneCore::UI
+
+  class GoogleServiceClient
+    def self.make_from_config(params: nil)
+      unless params[:json_key] || params[:json_key_data]
+        UI.important("To not be asked about this value, you can specify it using 'json_key'")
+        params[:json_key] = UI.input("The service account json file used to authenticate with Google: ")
+      end
+
+      if params[:json_key]
+        service_account_json = File.open(File.expand_path(params[:json_key]))
+      elsif params[:json_key_data]
+        service_account_json = StringIO.new(params[:json_key_data])
+      end
+
+      return self.new(service_account_json: service_account_json, params: params)
+    end
+
+    # Initializes the service and its auth_client using the specified information
+    # @param service_account_json: The raw service account Json data
+    def initialize(service_account_json: nil, params: nil)
+      scope = @scope
```

Comment:
`AbstractGoogleServiceClient` sounds good - the base class should not be usable by itself as it is missing the scope and service config.

`@scope` and `@service` being set in `initialize` is an ugly workaround.  

What I actually wanted to have is a base class and two implementations for it, that are "configured" differently. Just using one should have the correct service and scope, using the other would have its own scope and service set.

I tried to set `@scope = 'foo'` on the class level but that didn't work, so I (temp) moved it into `initialize`, which leads to all those problems.

How can I achieve this?

Reply:
I think maybe you want something like this? https://stackoverflow.com/a/19382692/2464643

```rb
class Foo
  TEST = "a"

  def speak
    puts self.class::TEST
  end
end

class Bar < Foo
  TEST = "b"
end

Bar.new.speak # => a
Foo.new.speak # => b
```

Reply:
I saw that, but it looked to strange at the moment.
But yep, that could work. What value would those have in my abstract class? Be empty? Default of the Supply::Client there as well?

Reply:
They would probably be `nil` in your abstract class and the you would probably need to just add an extra check/verification in your abstract class to make sure they are set because otherwise 💥 

This is a bit weird which is why the approach I would probably go with is passing them into the initialize parameter as this feel like a dependency injection thing. Passing them into the initialize parameter makes the most sense to me since the abstract class doesn't have any of its own so it feels weird to define a class level variable for them. 🙃

Reply:
So they would be defined and stored in `make_from_config` and then given to `new` as a param? (That feels pretty wrong as well)

Reply:
At least it will be obvious and pretty clear.

Reply:
Both `AbstractGoogleServiceClient` and `self.class::SERVICE`/`self.class::SCOPE` are now used.